### PR TITLE
Mapreduce style group-by aggregators 

### PR DESCRIPTION
**Pull request by nirandaperera**

### CODE REVIEW

File: cpp/src/cylon/mapreduce/mapreduce.cpp [hunk 1]
```diff
@@ -0,0 +1,1033 @@
...
+  *output = arrow::Table::Make(std::move(out_schema), std::move(out_arrays));
+  return Status::OK();
+}
+
+// todo test this with new kernels
+Status ShuffleTable(const std::shared_ptr<CylonContext> &ctx,
+                    const std::shared_ptr<arrow::Table> &atable,
+                    const std::vector<int> &key_cols, const AggKernelVector &agg_kernels,
+                    std::shared_ptr<Table> *shuffled, std::vector<int> *new_key_cols,
+                    AggKernelVector *new_agg_kernels) {
```
Comment (chathurawidanage) [inline cpp/src/cylon/mapreduce/mapreduce.cpp:830]: const?

File: cpp/src/cylon/mapreduce/mapreduce.cpp [hunk 2]
```diff
@@ -0,0 +1,1033 @@
...
+  static T Call(const T &x, const T &y) { return std::min(x, y); };
+};
+template<typename ArrowT>
+struct MinKernelImpl : public MapReduceKernelImpl1D<ArrowT, MinFunction> {
+  std::string name() const override { return "min"; }
+};
+
+template<typename T>
+struct MaxFunction {
+  static T Call(const T &x, const T &y) { return std::min(x, y); };
```
Comment (chathurawidanage) [inline cpp/src/cylon/mapreduce/mapreduce.cpp:122]: std::max

File: cpp/src/cylon/mapreduce/mapreduce.cpp [hunk 3]
```diff
@@ -0,0 +1,1033 @@
...
+
+#include <arrow/buffer_builder.h>
+#include <arrow/visitor_inline.h>
+
+namespace cylon {
+namespace mapred {
+
+template<typename ArrowT, typename Visitor>
+void CombineVisit(const std::shared_ptr<arrow::Array> &value_col, const int64_t *local_group_ids,
+                  Visitor &&visitor) {
```
Comment (chathurawidanage) [inline cpp/src/cylon/mapreduce/mapreduce.cpp:17]: Just asking out of curiosity... 

Would like to know whether the compiler inlines the visitor at 23. 
Would making visitor const give any compilation time advantage?

File: cpp/src/cylon/mapreduce/mapreduce.hpp [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+namespace cylon {
+namespace mapred {
+
+struct MapToGroupKernel {
+  virtual ~MapToGroupKernel() = default;
+  virtual Status Map(const arrow::ArrayVector &arrays,
+                     std::shared_ptr<arrow::Array> *local_group_ids,
+                     std::shared_ptr<arrow::Array> *local_group_indices,
+                     int64_t *local_num_groups,
+                     arrow::MemoryPool *pool) const;
```
Comment (chathurawidanage) [inline cpp/src/cylon/mapreduce/mapreduce.hpp:23]: missing  = arrow::default_memory_pool()

Comment (chathurawidanage) [reply cpp/src/cylon/mapreduce/mapreduce.hpp:23]: or should we avoid that in line 30?

Comment (nirandaperera) [reply cpp/src/cylon/mapreduce/mapreduce.hpp:23]: I think virtual functions can not have default values.

File: cpp/src/cylon/mapreduce/mapreduce.cpp (hunk 3 above)
Comment (nirandaperera) [reply cpp/src/cylon/mapreduce/mapreduce.cpp:17]: yes, from what I've seen previously, the compiler does inline that visitor call.

File: cpp/src/cylon/mapreduce/mapreduce.cpp (hunk 2 above)
Comment (nirandaperera) [reply cpp/src/cylon/mapreduce/mapreduce.cpp:122]: Thanks good catch! :-)

File: cpp/src/cylon/mapreduce/mapreduce.cpp (hunk 1 above)
Comment (nirandaperera) [reply cpp/src/cylon/mapreduce/mapreduce.cpp:830]: I didn't understand this :slightly_frowning_face:

**Review by chathurawidanage — APPROVED** [review]

**Pull request merged**